service/kernel/data_management/finiteness_checker.cpp

```diff
@@ -316,60 +491,56 @@ services::Status allValuesAreFiniteImpl(NumericTable & table, bool allowNaN, boo
     const size_t nElements = nRows * nColumns;
     const NTLayout layout  = table.getDataLayout();
 
-    size_t nDataPtrs, nElementsPerPtr;
     if (layout == NTLayout::soa)
     {
-        // SOA layout: pointer for each column
-        nDataPtrs       = nColumns;
-        nElementsPerPtr = nRows;
+        // first stage: compute sum of all values and check its finiteness
+        double sum       = computeSumSOA<cpu>(table);
```

Comment:
There is no need to rewrite code significantly and add specific function.
My concept of conversions removal:
1) Add type check from NumericTableDictionaryPtr
2) If type is applicable, get block of rows or column values
3) Pass pointer with data to function that will process it

`size_t nDataPtrs, nElementsPerPtr;`
`if (layout == NTLayout::soa)`
`{`
`        // SOA layout: pointer for each column`
`        nDataPtrs       = nColumns;`
`        nElementsPerPtr = nRows;`
`}`
`else`
`{`
`        // AOS layout: one pointer for all data`
`        nDataPtrs       = 1;`
`        nElementsPerPtr = nRows * nColumns;`
`}`
`NumericTableDictionaryPtr tableFeaturesDict = table.getDictionarySharedPtr();`
`for (size_t i = 0; i < nDataPtrs; ++i)`
`    switch ((*tableFeaturesDict)[j].getIndexType())`
`     case daal::data_management::features::IndexNumType::DAAL_FLOAT32:`
`          table.getBlockOfColumnValues()`
`          ...`
`          sum = computeSum(dataPtr, ...)`